### ISSUE TITLE
Investigate implementation of Electron's crash reporter API

### ISSUE THREAD

**Issue by hermwong** [phonegap/phonegap-app-desktop#607]
http://electron.atom.io/docs/v0.37.6/api/crash-reporter/

**Comment by hermwong** [comment]
Add analytics to crashes / failures, so that we can track how often the Desktop App crashes.

**Comment by hermwong** [comment]
Starting to make some progress.

I deployed a test server to [Heroku](https://fathomless-anchorage-12478.herokuapp.com/) using the following [project](https://github.com/johnmuhl/electron-crash-report-server).

Using a proxy, I was able to trace that the Desktop App is attempting to post the crash report to the server.

<img width="1017" alt="screen shot 2017-03-14 at 3 14 38 pm" src="https://cloud.githubusercontent.com/assets/107754/23924827/8d748612-08c9-11e7-9395-df879ece45da.png">

Currently looking into 500 error returned by the server.

**Comment by hermwong** [comment]
The other thing to note is that the `crashReporter` API generates the crash report in a `.dmp` file which we can't modify / parse within the Electron App.

**Comment by hermwong** [comment]
Was able to get the `crashReporter` to successfully `POST` to the Heroku server.

<img width="651" alt="screen shot 2017-03-16 at 10 59 07 am" src="https://cloud.githubusercontent.com/assets/107754/24011296/a1616d9c-0a37-11e7-8540-248f360a7416.png">

The server will list the crash reports for us and allow us to download the `.dmp` files. The next step is to see if we can find a way to read info with the `.dmp` files.

**Comment by hermwong** [comment]
More info on [`.dmp` file format](https://fileinfo.com/extension/dmp).

**Comment by hermwong** [comment]
See if we can use the `crashReporter.start` API to use the `extra` property to store a JSON object with more additional info:

- userID
- OS version
- more info regarding the error

Modify crash reporter server to trigger `crashEvent` with JSON meta data so that we can track crashes in the analytics server.

Modify the crash reporter server to disable the storage of DUMP files (we may re-enable this feature in the future).

**Comment by hermwong** [comment]
I downloaded & installed Visual Studio 2017 on my Win10 VM to view the `.dmp` file. Here is an example of the data we can find within the `.dmp` file.

<img width="1049" alt="screen shot 2017-03-20 at 12 16 10 pm" src="https://cloud.githubusercontent.com/assets/107754/24117362/2a553a5c-0d67-11e7-89c6-b6aad0dc614f.png">

**Comment by hermwong** [comment]
Got the Desktop App to post of all the necessary data with the Crash Reporter to the crash reporter server.

The crash reporter server now receives the JSON data and formats it into the proper GELF JSON format for our analytics server.

<img width="822" alt="screen shot 2017-03-21 at 3 04 24 pm" src="https://cloud.githubusercontent.com/assets/107754/24173096/fbc66922-0e47-11e7-8811-5c6ba83f9a88.png">

Next step is to get the crash reporter server to send the GELF JSON over to the analytics server.

**Comment by hermwong** [comment]
@filmaj pointed out that the IP information that the analytics server will try to infer from the request will come from the crash reporter server, and not from the client.

We will need to see if the Desktop App can include the client IP as part of the JSON data being passed out in the crashReporter's `extra` property.

The crash reporter server should POST to the URL `https://metrics.phonegap.com/gelfproxypass` to prevent the IP inferring by the analytics server.

**Comment by filmaj** [comment]
Furthermore to your point @hermwong, please try to set a top-level JSON property `ip` to be a string containing the client's IP.

**Comment by filmaj** [comment]
I've added information about the `gelfproxypass` route on the metrics server, as well as information about what kind of IP-related fields will be automatically added to events, to the metrics server readme: https://github.com/phonegap/phonegap-metrics

**Comment by hermwong** [comment]
When trying to retrieve the client IP from the Desktop App, I was capturing an internal IP address assigned to my computer by my router.

I found the best way to capture the client's external IP address was to grab the client IP address on the crash reporter's server from the request object.

<img width="432" alt="screen shot 2017-03-22 at 10 54 01 am" src="https://cloud.githubusercontent.com/assets/107754/24213021/66487726-0eee-11e7-9427-3f605e111744.png">

**Comment by filmaj** [comment]
nice!

**Comment by hermwong** [comment]
First piece of test data into the analytics server from crash reporter!

<img width="1371" alt="screen shot 2017-03-22 at 1 23 26 pm" src="https://cloud.githubusercontent.com/assets/107754/24219065/c755343c-0f02-11e7-810e-877a8d2ae053.png">

**Comment by hermwong** [comment]
TODO Checklist before we can release this feature to production.

1. transfer ownership of the crash reporter server to PhoneGap GH, it is currently located here: https://github.com/hermwong/electron-crash-report-server
2. migrate the heroku server to the PhoneGap heroku account (if we have one?)
3. do we want to change the name of the crash reporter server's url? It's currently `https://fathomless-anchorage-12478.herokuapp.com/` this would also require us to update the url in the Desktop App's codebase.

**Comment by hermwong** [comment]
Finished modifying the server so that it no longer writes the `.dmp` files and also removed the `download dump file` button from the crash reports.

**Comment by filmaj** [comment]
I believe the phonegap heroku account is maxed out for its free tier 😮

**Comment by hermwong** [comment]
The crash reporter data we’re currently capturing is from Electron’s `renderer process`, I’ve been looking into capturing crashes in Electron’s `main process`.

Electron’s `main process` is responsible for bootstrapping the Electron App, while the `renderer process` is responsible for the content within the browser inside the Electron App.

I’ve discovered that the `main process` doesn’t have access to WebAPIs (only the `renderer process` has access to WebAPIs, which means we’re not able to access the `localStorage` (where we store the `userID`).

I can use the `IPCRenderer` to `IPCMain` communication bridge to push all of the necessary data to the `main process` but only after the `renderer process` has finished rendering the UI.

The only use case where we aren’t able to capture the `userID` is if the `main process` has a crash before the `renderer process` has finished rendering the UI. 

In this scenario should we:

1. just use a blank `userID` or some placeholder value when recording crashes from the `main process`?
2. ignore this use case?

**Comment by mwbrooks** [comment]
Good summary @hermwong.

**Should we capture crashes in `main process`?**

I think we should. These sound like serious crashes when the main process cannot start up.

**Should we capture the `userID` when the `main process` crashes?**

The `userID` would be helpful for querying unique crashes. If we see 10 `main process` crashes happen, then a `userID` would help us understand if it's 1 user or 10 users.

Personally, I don't think we can rely on pushing data from `renderer process` to `main process`. I suspect that some of the `main process` crashes will happen on launch before the `renderer process` starts.

The two options that come to mind are:

1. Offline @purplecabbage suggested reading/writing to a filesystem configuration file rather than using `localStorage` for the `userID`. There are a lot of positives around persistence but today we use `localStorage`. If it's trivial to migrate to the file system storage, then that's our best solution.
1. Use a blank `userID` to record `main process` crashes. If we can attach some UUID to the report, then that would serve the same purpose as the `userID`. Then in the re-write, we want to ensure that we migrate to file system storage.

**Comment by filmaj** [comment]
Hey @hermwong, just based on some work on updating the analytics in the phonegap-cli, I've added some more info to https://github.com/phonegap/phonegap-metrics. In particular, what I think may be applicable for this issue for the desktop app:

 - if you can track the node runtime version in a `_nodeVersion` property that'd be pretty cool.
 - for the `_platform` property, if possible, could we use the [`os-name`](npmjs.com/package/os-name) node module to set its value? The CLI already uses that in GA, and it'd be nice to have parity across products.

**Comment by hermwong** [comment]
Currently looking into storing the `userID` in a file rather than `localStorage`.

I'll also add the `_nodeVersion` property and update the `_platform` property while working on the `userID` storage fix.

**Comment by filmaj** [comment]
Sweet, here's some more info on how the phonegap-cli extracts or creates a user ID for use in Google Analytics:

1. Firstly, the library that the CLI uses to handle interacting with GA is called [insight](https://www.npmjs.com/package/insight). [Under the hood, it leverages](https://github.com/yeoman/insight/blob/master/lib/index.js#L38) a module called [Configstore](https://www.npmjs.com/package/configstore). Configstore is basically a key/value store that's persisted on the filesystem. Configstore stores things (on Unix-like environments) under the user's home directory in a hidden `.config` directory. So for example, on a UNIX system, the phonegap-cli stores its analytics-related configuration under `~/.config/configstore/insight-phonegap.json`. By default in `insight`, the filename of the config file is set to `insight-<packagename>` - thus in the phonegap-cli, it is`insight-phonegap`. Furthermore, [the key within this JSON file that `insight` uses to store the user id is called `clientId`](https://github.com/yeoman/insight/blob/master/lib/index.js#L50-L52).
2. The phonegap-cli, moving forward, will leverage this Configstore module and use the same filename + config key to retrieve the user ID, when setting user IDs. It should be as easy as:

        var conf = new Configstore('insight-phonegap');
        var key = 'clientId';
        if (!conf.has(key)) {
             conf.set(key, 'generate a new user id here');
        }
        var userID = conf.get(key);

3. The cool thing about using this approach is that if more of our products consolidate on this, then our analytics would be using the same user ID across products.

**Comment by hermwong** [comment]
Using [`configstore`](https://www.npmjs.com/package/configstore), I'm able to write a configuration file to store data on the filesystem.

As a quick proof of concept, I was able to write `pgdesktop.json` to the `/Users/<username>/.config/configstore` directory (on OSX). I was able to set & get a value from the `pgdesktop.json` file using the `configstore` npm package.

**Comment by filmaj** [comment]
Wait wait! But but! Can you change `pgdesktop.json` to `insight-phonegap.json` ? This way if a user installed phonegap-cli on their machine before they installed the desktop app (or vice-versa) we'd track the same user ID. Also important is to leverage the `clientId` key in said configstore. This way we're backwards compatible and we have the option of tracking the same user across different products on the same machine!

**Comment by hermwong** [comment]
Yeap, easy to change it. I was just using the `pgdesktop.json` as a quick test to see if I could get a quick working example of setting & getting a value from the `configstore`.

**Comment by hermwong** [comment]
Currently have the `insight` package initializing with our `trackingCode`, however the `if` condition doesn't seem to work properly and keeps allowing the app to overwrite the `clientId`.

```
var pkg = require('./package.json');

const Insight = require('insight');
const insight = new Insight({
    trackingCode: 'UA-94271-31',
    pkg
});

    // TODO: fix this check - seems to keep going into the condition & overwriting the clientId!!
    if (!insight.clientId) {
        insight.clientId(generateId());
    }
```

**Comment by filmaj** [comment]
What `pkg` did you provide to Insight?

**Comment by hermwong** [comment]
Was testing using `var pkg = require('./package.json');` which just loads the PG-Desktop's `package.json`. 

It's possible that I need to modify it to point to the `insight-phonegap.json` file; which would mean that we'd have to determine the path to `insight-phonegap.json` depending on the user's OS (Windows/OSX).

**Comment by filmaj** [comment]
[`Insight` uses a string in the form "insight-<pkg.name>", "pkg" being the package.json of the parent module as a unique configstore ID](https://github.com/yeoman/insight/blob/master/lib/index.js#L38). It then passes that ID to Configstore as the filename in the Configstore constructor:

    new Configstore(`insight-${this.packageName}`, {
        clientId: options.clientId || Math.floor(Date.now() * Math.random())
    });

That `this.packageName` is based on the parent module. So in phonegap-cli's case, that would be equal to "insight-phonegap". However, in the Desktop app's case, it would resolve to "insight-<whatever-`name`-is-in-desktop's-package.json>". That would be maybe "insight-PhoneGap" in [the desktop app's case](https://github.com/phonegap/phonegap-app-desktop/blob/master/package.json#L2). Perhaps the casing is different? You can check what Configstores you have on your machine by checking the contents of `~/.config/configstore/`. 

Try not using Insight at all, and use the Configstore object directly. I think if we pass the string "insight-phonegap" to the Configstore constructor, under the hood Configstore will resolve to the same file:

    var conf = new Configstore('insight-phonegap');

[Based on Configstore's path resolution logic in its constructor](https://github.com/yeoman/configstore/blob/master/index.js#L17-L25), using the same `id` parameter gives us a pretty good chance of getting the same file.

**Comment by hermwong** [comment]
After a bit more investigation, it seems that the casing is causing the `.json` to be overwritten. The `cli` installs `insight-phonegap.json` and then generates `cliendId=X`. When the Desktop App runs it installs `insight-Phonegap.json` which ends up overwriting the `.json` file created by the `cli` and results in a new `clientId` being generated.

**Comment by hermwong** [comment]
Got the App retrieving / setting the `userId` as `clientId` in the `insight-phonegap.json`.

Thinking that we need move the "Send analytics" setting into `insight-phonegap.json` from `localStorage`.

**Comment by purplecabbage** [comment]
Cool!
'Send analytics' means the user has chosen to opt in?

**Comment by hermwong** [comment]
Yeap - our actual copy in the Desktop App reads "Send anonymous diagnostic & usage data" next to a checkbox.

**Comment by hermwong** [comment]
I think the best workflow to handle the `sendAnalytics` flag would be:

check if `sendAnalytics` exists in `insight-phonegap.json`
  - if it exists, retrieve the value and use it throughout the Desktop App session
  - else, retrieve `sendAnalytics from `localStorage` (this is where the current flag is stored) and write the value into `insight-phonegap.json`

**Comment by filmaj** [comment]
Slight derail question here, @hermwong, but is there anyway we can log unique sessions in the Desktop app? Something like, generating an ID when the app starts, and logging that ID in a `_session` field?

I've documented the `_session` field up on github.com/phonegap/phonegap-metrics.

**Comment by hermwong** [comment]
I can generate a guid each time a desktop session is started and log that as a `_session` property.

**Comment by hermwong** [comment]
Next few modifications required:

- use `os-name` package for the `_platform` property
- add `_nodeVersion` property
- add `_session` property
- update the `port` in settings to use `configstore` instead of `localStorage`

**Comment by hermwong** [comment]
Result of a quick test with all the new fields added to analytics.

<img width="305" alt="screen shot 2017-03-30 at 2 50 10 pm" src="https://cloud.githubusercontent.com/assets/107754/24527696/33bd03ec-1558-11e7-9e43-7afe970226e4.png">

**Comment by filmaj** [comment]
So freaking awesome! Nice job! I think that about covers it for GA parity!

**Comment by hermwong** [comment]
During testing - discovered that crashreporter was having some issues retrieving `userID` and `session` from the `configstore`.

Fixed the issue in the `browser process` with this [commit](https://github.com/phonegap/phonegap-app-desktop/commit/ce5f02f2915262a9447f77dad0f1f852d58acee0).

Looking into the fix for the `renderer process`.

**Comment by surajpindoria** [comment]
I managed to get the crash reporter to send to the server in the renderer process. Also refactored how we set the debug flag, so it should be reported properly if there is a crash.

@hermwong if you can review my last commit, I changed how we instantiate the main window. It should work the same as before but just in case I missed something.

**Comment by hermwong** [comment]
@surajpindoria crash reporter stuff looks good. I was able to verify that the crash reporter was posting from the renderer and browser processes to the test server.

**Comment by hermwong** [comment]
Tested the production crash reporter server, it's collecting crash reporter data and posting analytics to metrics.phonegap.com.
